nemo/core/classes/modelPT.py

```diff
@@ -1831,24 +1864,50 @@ def on_train_batch_start(self, batch: Any, batch_idx: int, unused: int = 0) -> O
         https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#on-train-batch-start
         We use it here to enable nsys profiling and dynamic freezing.
         """
-
+        # logging.info(f"Training Real batch {self._real_batch_idx} started.")
         # nsys profiling
         if self.device.type == 'cuda':
             if hasattr(self, '_nsys_profile_enabled'):
                 if self._nsys_profile_enabled and not self._nsys_profile_started:
-                    if batch_idx >= self._nsys_profile_start_step and get_rank() in self._nsys_profile_ranks:
+                    if self._real_batch_idx >= self._nsys_profile_start_step and get_rank() in self._nsys_profile_ranks:
                         logging.info("====== Start nsys profiling ======")
                         torch.cuda.cudart().cudaProfilerStart()
                         if self._nsys_profile_gen_shape:
                             torch.autograd.profiler.emit_nvtx(record_shapes=True).__enter__()
                         self._nsys_profile_started = True
 
             if hasattr(self, '_memory_profile_enabled'):
+                if self._memory_profile_enabled:
+                    if self._memory_profile_weight_enabled and (self._real_batch_idx == 0) and (get_rank() == self._memory_profile_rank): # before batch-0 start
+                        try:
+                            logging.info(f"====== Rank[{self._memory_profile_rank}], Batch[{batch_idx}]. End CUDA memory profiling: Weight. Snapshot saved in {self._memory_profile_snapshot_file_weight} ======")
+                            torch.cuda.memory._dump_snapshot(f"{self._memory_profile_snapshot_file_weight}")
+                            # torch.cuda.memory._record_memory_history(enabled=None)
+                        except Exception as e:
+                            logging.error(f"Failed to capture memory snapshot {e}")
+                            return
+
+                        # Call the analysis function
+                        if self._memory_profile_analysis_enabled:
```

Comment:
do you need this if ? I would assume `_memory_profile_analysis_enabled` does not change value and in line 1880 you already check whether it's true or not.